063-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,208 @@
+    ```
+    This is due these [mapping rules](https://github.com/strimzi/strimzi-kafka-operator/blob/main/examples/metrics/kafka-metrics.yaml#L127-L135). 
+
+- The `count` suffix is used instead of `total` for some Counters by jmx_exporter:
+    ```
+    kafka_network_acceptor_acceptorblockedpercent_count{listener="PLAINTEXT",} # jmx_exporter
+    kafka_network_acceptor_acceptorblockedpercent_total{listener="PLAINTEXT",} # reporter
+    ```
+    Prometheus recommends using the `total` suffix and it's actually automatically injected by the Java library. The example jmx_exporter rules correctly replace the suffix in most metrics (like in the metrics mentioned above with `PerSec`) but they don't match all metrics hence the inconsistency.
+
+- Some labels are wrapped twice in quotes by jmx_exporter
+    ```
+    # jmx_exporter
+    kafka_log_logcleanermanager_uncleanable_bytes{logDirectory="\"/tmp/kafka-logs\"",}
+    kafka_log_logcleanermanager_uncleanable_partitions_count{logDirectory="\"/tmp/kafka-logs\"",}
+    kafka_log_logmanager_logdirectoryoffline{logDirectory="\"/tmp/kafka-logs\"",}
+    # reporter
+    kafka_log_logcleanermanager_uncleanable_bytes{logDirectory="/tmp/kafka-logs",}
+    kafka_log_logcleanermanager_uncleanable_partitions_count{logDirectory="/tmp/kafka-logs",}
+    kafka_log_logmanager_logdirectoryoffline{logDirectory="/tmp/kafka-logs",}
+    ```
+    I'm guessing this is done to support paths containing special characters. In the context of Strimzi I don't think this is necessary.
+
+
+In terms of performance, in my very limited testing the reporter is much faster than jmx_exporter. This is not a benchmark or rigorous by any means, I've simply been running `time curl --head http://localhost:{PORT}/metrics` against a single broker on my laptop with both the reporter (on port 8080) and jmx_exporter (on port 9090) running. Since each partition has metrics, I used the number of partition as a proxy to increase the total number of metrics.
+
+| # of partitions | jmx_exporter | reporter |
+| --- | --- | --- |
+| 1 | 600ms | 100ms |
+| 500 | 900ms | 300ms |
+| 1000 | 1300ms | 400ms |
+| 2000 | 2000ms | 800ms | 
+
+## Rejected alternatives
+
+- I considered using [Micrometer](https://micrometer.io/) in the reporter. The benefit is that it would allow exporting metrics to different monitoring systems. The issue is that it requires converting the Kafka and Yammer metrics into the Micrometer format and then have Micrometer export that to Prometheus. As each metric library has its own characteristics, chaining several conversions can lead to slightly different semantics. Finally Prometheus seems to be the leading monitoring solution in the Kubernetes ecosystem, and I expect most other monitoring tools to integrate this it. For these reasons I decided to not use Micrometer. Ideally Kafka would export its metrics via Micrometer. It's something I've started to explore but it is definitively a very difficult task.
```

Comment:
```suggestion
- I considered using [Micrometer](https://micrometer.io/) in the reporter. The benefit is that it would allow exporting metrics to different monitoring systems. The issue is that it requires converting the Kafka and Yammer metrics into the Micrometer format and then have Micrometer export that to Prometheus. As each metric library has its own characteristics, chaining several conversions can lead to slightly different semantics. Finally Prometheus seems to be the leading monitoring solution in the Kubernetes ecosystem, and I expect most other monitoring tools to integrate this way. For these reasons I decided to not use Micrometer. Ideally Kafka would export its metrics via Micrometer. It's something I've started to explore but it is definitively a very difficult task.
```

Reply:
I meant to write `integrate with it`. I'll push an update.

Reply:
Done